KAFKA-18276 Migrate ProducerRebootstrapTest to new test infra

clarkwtc opened the pull request:
The PR changed three things.
* Migrated `ProducerRebootstrapTest` to new test infra and removed the old Scala test.
* Updated the original test case to cover rebootstrap scenarios.
* Integrated `ProducerRebootstrapTest` into `ClientRebootstrapTest` in the `client-integration-tests` module.

Default `ProducerRebootstrap` config:
> properties.put(CommonClientConfigs.METADATA_RECOVERY_STRATEGY_CONFIG, "rebootstrap");
       properties.put(CommonClientConfigs.METADATA_RECOVERY_REBOOTSTRAP_TRIGGER_MS_CONFIG, "300000");
       properties.put(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG, "10000");
       properties.put(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG, "30000");
       properties.put(CommonClientConfigs.RECONNECT_BACKOFF_MS_CONFIG, "50L");
       properties.put(CommonClientConfigs.RECONNECT_BACKOFF_MAX_MS_CONFIG, "1000L");
       
The test case for the producer with enabled rebootstrap
<img width="1549" alt="Screenshot 2025-03-17 at 10 46 03 PM" src="https://github.com/user-attachments/assets/547840a6-d79d-4db4-98c0-9b05ed04cf60" />

The test case for the producer with disabled rebootstrap
<img width="1552" alt="Screenshot 2025-03-17 at 10 46 47 PM" src="https://github.com/user-attachments/assets/2248e809-d9d5-4f3b-a24f-ba1aa0fef728" />

Reviewers: TengYao Chi <kitingiao@gmail.com>, Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

Commit:
feat: Migrate RebootstrapTest to new test infra

Commit:
feat: Remove old `ProducerRebootstrapTest` test

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+    public void testRebootstrap(ClusterInstance clusterInstance) throws ExecutionException, InterruptedException {
+        String topic = "topic";
+        try (Admin admin = clusterInstance.admin()) {
+            admin.createTopics(Collections.singletonList(new NewTopic(topic, BROKER_COUNT, (short) 2))).all().get();
```
Please use `List.of()` instead `Collections.singletonList`, and  I don't think we need to get the future value.
```
admin.createTopics(List.of(new NewTopic(topic, BROKER_COUNT, (short) 2)));
```

clarkwtc replied:
Thank you for your comments.
Sure, we don't need to get any future value.

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+    @ClusterTemplate(value = "generator")
+    public void testRebootstrap(ClusterInstance clusterInstance) throws ExecutionException, InterruptedException {
+        String topic = "topic";
+        try (Admin admin = clusterInstance.admin()) {
```
nit: Variables should be declared using `var` to maintain consistency throughout the file.

clarkwtc replied:
I've fixed it to maintain consistency by using `var`.

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+
+        try (var producer = clusterInstance.producer()) {
+            // Only the server 0 is available for the producer during the bootstrap.
+            RecordMetadata recordMetadata0 = producer.send(new ProducerRecord<>(topic, part, "key 1".getBytes(), "value 1".getBytes())).get();
```
ditto: Variables should be declared using var to maintain consistency throughout the file.

clarkwtc replied:
I've fixed it to maintain consistency by using `var`.

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+            // Current server 0 is offline.
+            // However, the server 1 from the bootstrap list is online.
+            // Should be able to produce records.
+            RecordMetadata recordMetadata1 = producer.send(new ProducerRecord<>(topic, part, "key 1".getBytes(), "value 1".getBytes())).get();
```
ditto: Variables should be declared using var to maintain consistency throughout the file.

clarkwtc replied:
I've fixed it to maintain consistency by using `var`.

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+            server0.startup();
+
+            // The same situation, but the server 1 has gone and server 0 is back.
+            RecordMetadata recordMetadata2 = producer.send(new ProducerRecord<>(topic, part, "key 1".getBytes(), "value 1".getBytes())).get();
```
ditto: Variables should be declared using var to maintain consistency throughout the file.

clarkwtc replied:
I've fixed it to maintain consistency by using `var`.

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        int part = 0;
+        KafkaBroker server0 = clusterInstance.brokers().get(0);
+        KafkaBroker server1 = clusterInstance.brokers().get(1);
```
ditto: Variables should be declared using var to maintain consistency throughout the file.

clarkwtc replied:
I've fixed it to maintain consistency by using `var`.

m1a2st commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,121 @@
+    static Map<String, String> getRebootstrapConfig(boolean useRebootstrapTriggerMs) {
+        Map<String, String> properties = new HashMap<>();
+        if (useRebootstrapTriggerMs) {
+            properties.put("metadata.recovery.rebootstrap.trigger.ms", "5000");
+        } else {
+            properties.put("metadata.recovery.rebootstrap.trigger.ms", "3600000");
+            properties.put("socket.connection.setup.timeout.ms", "5000");
+            properties.put("socket.connection.setup.timeout.max.ms", "5000");
+            properties.put("reconnect.backoff.ms", "1000");
+            properties.put("reconnect.backoff.max.ms", "1000");
+        }
+        properties.put("metadata.recovery.strategy", "rebootstrap");
```
We could use `CommonClientConfigs.METADATA_RECOVERY_REBOOTSTRAP_TRIGGER_MS_CONFIG` instead of plaintext, other configs are same.

clarkwtc replied:
I can't do it because the styleAction check blocks the reference by`CommonClientConfigs` in tests.
I only just wrote it in plaintext.

m1a2st replied:
You can relax the conditions for core module `import-control`.

clarkwtc replied:
Thank you, the method effectively relaxes the conditions.
I have tested to update the `org.apache.kafka.clients` pkg in `import-control-core.xml`. 
Is it okay to submit this change directly on commit?

m1a2st replied:
I think it's not a problem.

m1a2st reviewed (commented):
Thanks for @clarkwtc, left some comments.

Commit:
fix: Change list object from Collections.singletonList to `List.of`

Reason: `List.of` optimizing for small size and catch bugs easier.

Commit:
style: Maintain consistency throughout the file by using var to replace declared variables

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

Commit:
chore: Remove unused code.

Commit:
feat: Relax `CommonClientConfigs` to core module by import-control

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

frankvicky commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,118 @@
+    private static final int BROKER_COUNT = 2;
+
+    static List<ClusterConfig> generator() {
+        Map<String, String> serverProperties = new HashMap<>();
+        // Enable unclean leader election for the test topic
+        serverProperties.put(TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, "true");
+        serverProperties.put(GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, String.valueOf(BROKER_COUNT));
```
nit: `Map.of()`

clarkwtc replied:
Thank you for your comments.
It is preferable to use an immutable map here.

frankvicky commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,118 @@
+public class ProducerRebootstrapTest {
+    private static final int BROKER_COUNT = 2;
+
+    static List<ClusterConfig> generator() {
```
`private`

clarkwtc replied:
I've fixed the access modifiers.

frankvicky commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,118 @@
+                .toList();
+    }
+
+    static Map<String, String> getRebootstrapConfig(boolean useRebootstrapTriggerMs) {
```
`private`

clarkwtc replied:
I've fixed the access modifiers.

frankvicky commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,118 @@
+
+        try (var producer = clusterInstance.producer()) {
+            // Only the server 0 is available for the producer during the bootstrap.
+            var recordMetadata0 = producer.send(new ProducerRecord<>(topic, part, "key 1".getBytes(), "value 1".getBytes())).get();
```
Please extract `key 1` and `value 1` as variables.

clarkwtc replied:
I've fixed it, then I found a mistake for sending the record with `key 0` and `value 0` first, when comparing it to the old Scala test.

frankvicky commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,118 @@
+        return properties;
+    }
+
+    static ClusterConfig buildConfig(Map<String, String> serverProperties, Map<String, String> rebootstrapProperties) {
```
`private`

clarkwtc replied:
I've fixed the access modifiers.

frankvicky reviewed (commented):
Thanks for the patch.
I have a few comments.

Commit:
refactor: Use an immutable map to avoid be modified and improve performance.

Commit:
refactor: Modify the access modifiers of functions.

Commit:
fix: Send record with `key 0` and `value 0` first.

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

Commit:
Fix: Remove unnecessary whitespace

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

m1a2st reviewed (approved):
Thanks for this PR, LGTM

chia7712 commented on core/src/test/java/kafka/test/api/ProducerRebootstrapTest.java:
```diff
@@ -0,0 +1,123 @@
+    public void testRebootstrap(ClusterInstance clusterInstance) throws ExecutionException, InterruptedException {
+        var topic = "topic";
+        try (var admin = clusterInstance.admin()) {
+            admin.createTopics(List.of(new NewTopic(topic, BROKER_COUNT, (short) 2)));
```
Should we use `BROKER_COUNT` to define `replicationFactor` rather than partition?

chia7712 commented:
this PR has same issues to https://github.com/apache/kafka/pull/19094#discussion_r1986208718

ijuma commented:
We should probably create the clients integration test module before doing these conversions so that the Java classes end up in a Java only module.

chia7712 commented:
> We should probably create the clients integration test module before doing these conversions so that the Java classes end up in a Java only module.

yes, @mumrah has created pull request #19144 to address this issue. After merging #19144, we can submit a minor pull request to relocate the Java test code from the core module to the client module.

Commit:
feat: Rewrite ProducerRebootstrapTest to cover the current scenario

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

clarkwtc commented:
@chia7712 
Thank you for your note.
I have also updated this PR to fix the related issues from https://github.com/apache/kafka/pull/19094#discussion_r1986208718

chia7712 commented:
@clarkwtc could you please move this test to clients-integration-tests ?

Commit:
feat: Move this test to clients-integration-tests

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

clarkwtc commented:
@chia7712 @ijuma 
I already moved this test to clients-integration-tests.
I have also added `<allow pkg="org.junit" />` on `import-control-clients-integration-tests.xml`

Commit:
chore: Make it better

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

Commit:
chore: Make it better

- Change Assertions to static import
- Change brokers variables to defined PARTITIONS

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

Commit:
fix: Redefine test case to use only one partitions.

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test

Commit:
feat: Move this test to clients-integration-tests

Commit:
chore: Make ProducerRecord be easier with one partition.

Commit:
Merge branch 'trunk' into KAFKA-18276-rewrite-producer_rebootstrap_test
